pkg/geo/geomfn/mvtgeom.go

```diff
@@ -0,0 +1,351 @@
+	// Don't check GeometryCollection so that we don't discard a collection of points.
+	case *geom.LineString, *geom.MultiLineString, *geom.Polygon, *geom.MultiPolygon:
+		geometryBoundingBox := gt.Bounds()
+		geomWidth := geometryBoundingBox.Max(0) - geometryBoundingBox.Min(0)
+		geomHeight := geometryBoundingBox.Max(1) - geometryBoundingBox.Min(1)
+		// We use half of the square height and width as limit instead of area so it works properly with lines.
+		halfBoundsWidth := ((bbox.HiX - bbox.LoX) / float64(extent)) / 2
+		halfBoundsHeight := ((bbox.HiY - bbox.LoY) / float64(extent)) / 2
+		if geomWidth < halfBoundsWidth && geomHeight < halfBoundsHeight {
+			return true
+		}
+	}
+	return false
+}
+
+// transformToMvtGeom transforms a geometry into vector tile coordinate space.
+func transformToMvtGeom(
+	gt geom.T, g geo.Geometry, bbox geopb.BoundingBox, extent int, buffer int, clipGeometry bool,
+) (*geo.Geometry, error) {
+	basicType, err := getBasicType(gt)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get geom.T basic type")
+	}
+	basicTypeGeometry, err := convertToBasicType(g, basicType)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to convert geometry to basic type")
+	}
+	if basicTypeGeometry.Empty() {
+		return &geo.Geometry{}, nil
+	}
+
+	removeRepeatedPointsGeometry, err := RemoveRepeatedPoints(basicTypeGeometry, 0)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to remove repeated points")
+	}
+	// Remove points on straight lines
+	simplifiedGeometry, empty, err := Simplify(removeRepeatedPointsGeometry, 0, false)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to simplify geometry")
+	}
+	if empty {
+		return &geo.Geometry{}, nil
+	}
+
+	affinedGeometry, err := transformToTileCoordinateSpace(simplifiedGeometry, bbox, extent)
+	if err != nil {
+		return nil, err
+	}
+
+	snappedGeometry, err := snapToIntegersGrid(affinedGeometry)
+	if err != nil {
+		return nil, err
+	}
+
+	out, err := clipAndValidateMvtOutput(snappedGeometry, basicType, extent, buffer, clipGeometry)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to clip and validate")
+	}
+	if out.Empty() {
+		return &geo.Geometry{}, nil
+	}
+	return &out, nil
+}
+
+func transformToTileCoordinateSpace(
+	g geo.Geometry, bbox geopb.BoundingBox, extent int,
+) (geo.Geometry, error) {
+	width := bbox.HiX - bbox.LoX
+	height := bbox.HiY - bbox.LoY
+	fx := float64(extent) / width
+	fy := -float64(extent) / height
+	affineMatrix := AffineMatrix{
+		{fx, 0, 0, -bbox.LoX * fx},
+		{0, fy, 0, -bbox.HiY * fy},
+		{0, 0, 1, 0},
+		{0, 0, 0, 1},
+	}
+	affinedGeometry, err := Affine(g, affineMatrix)
+	if err != nil {
+		return geo.Geometry{}, errors.Wrap(err, "failed to affine geometry")
+	}
+	return affinedGeometry, nil
+}
+
+// clipAndValidateMvtOutput prepares the geometry so that it's clipped (if needed), valid, snapped to grid and it's basic type is the same as the input geometry.
+func clipAndValidateMvtOutput(
```

Comment:
nit: rename to `clipAndValidateMVTOutput`

Reply:
done